### PR TITLE
webpack 5 compatibiltiy

### PR DESCRIPTION
**Pull request by jantimon**
Thanks to @PutziSan and @fivethreeo we have now a first working webpack 5 version!!

### CODE REVIEW

File: .github/workflows/nodejs.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+name: Node.js CI
```
Comment (jantimon) [inline .github/workflows/nodejs.yml:1]: can we please keep this in travis and migrate from travis to github actions in a later pr?

File: .github/workflows/npm_release.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
+name: Do release
```
Comment (jantimon) [inline .github/workflows/npm_release.yml:1]: can we please discuss about a github action release feature in another pull request and remove it for now?

File: lib/compiler.js [hunk 3]
```diff
@@ -288,63 +289,63 @@ function getFileDependencies (compiler) {
 }
 
 /**
- * @type {WeakMap<WebpackCompilation, WeakMap<HtmlWebpackChildCompiler, boolean>>}}
+ * @type {WeakMap<object, Snapshot>}
  */
-const hasOutdatedCompilationDependenciesMap = new WeakMap();
+const snapshotMap = new WeakMap();
+const fileDepependenciesKeyRef = {};
```
Comment (jantimon) [inline lib/compiler.js:295]: This looks really dangerous as it is one single global variable used for (potentially) multiple compilers, compilations and html-webpack-plugin instances.

File: lib/compiler.js [hunk 4]
```diff
@@ -288,63 +289,63 @@ function getFileDependencies (compiler) {
...
-  if (!hasOutdatedChildCompilerDependenciesMap) {
-    hasOutdatedChildCompilerDependenciesMap = new WeakMap();
-    hasOutdatedCompilationDependenciesMap.set(mainCompilation, hasOutdatedChildCompilerDependenciesMap);
+
+  const keyStr = JSON.stringify([...childCompiler.fileDependencies].sort());
+  // get reference for key-strings to use it with WeakMap:
+  let key = fileDepependenciesKeyRef[keyStr];
+  if (!key) {
+    key = {};
+    fileDepependenciesKeyRef[keyStr] = key;
```
Comment (jantimon) [inline lib/compiler.js:313]: This is a memory leak as it will grow on every compilation if output name contains a hash.

File: lib/compiler.js [hunk 5]
```diff
@@ -288,63 +289,63 @@ function getFileDependencies (compiler) {
...
-  /**
-   * @type {WeakMap<HtmlWebpackChildCompiler, boolean>|undefined}
-   */
-  let hasOutdatedChildCompilerDependenciesMap = hasOutdatedCompilationDependenciesMap.get(mainCompilation);
-  // Create map for childCompiler if none exist
-  if (!hasOutdatedChildCompilerDependenciesMap) {
-    hasOutdatedChildCompilerDependenciesMap = new WeakMap();
-    hasOutdatedCompilationDependenciesMap.set(mainCompilation, hasOutdatedChildCompilerDependenciesMap);
+
+  const keyStr = JSON.stringify([...childCompiler.fileDependencies].sort());
```
Comment (jantimon) [inline lib/compiler.js:308]: This is probably way slower than comparing the file dependencies

File: spec/caching.spec.js [hunk 6]
```diff
@@ -128,6 +139,7 @@ describe('HtmlWebpackPluginCaching', () => {
       .then(stats => {
         // Expect no errors:
         expectNoErrors(stats);
+        console.log(getCompiledModuless(stats.toJson()));
```
Comment (jantimon) [inline spec/caching.spec.js:142]: was that added for debugging only?

File: spec/hot.spec.js [hunk 7]
```diff
@@ -81,8 +85,12 @@ describe('HtmlWebpackPluginHMR', () => {
         const hotUpdateJsFileNames = Object.keys(stats.compilation.assets).filter((fileName) => /\.hot-update\.js$/.test(fileName));
         expect(hotUpdateJsFileNames).not.toEqual([]);
         expect(hotUpdateJsFileNames.length).toEqual(1);
-        const hotUpdateFileSource = stats.compilation.assets[hotUpdateJsFileNames[0]].source();
-        expect(hotUpdateFileSource).not.toEqual('');
+        if (webpackMajorVersion < 5) {
```
Comment (jantimon) [inline spec/hot.spec.js:88]: Why do we need this check?

What about removing the old `hotUpdateFileSource` code as you added tests based on size anyway?

File: lib/compiler.js [hunk 8]
```diff
@@ -288,63 +289,63 @@ function getFileDependencies (compiler) {
...
-}
 
-/**
- * Returns `true` if the file dependencies of the given childCompiler are outdated.
- *
- * @param {WebpackCompilation} mainCompilation
- * @param {HtmlWebpackChildCompiler} childCompiler
- * @returns {boolean}
- */
-function isChildCompilerCacheOutdated (mainCompilation, childCompiler) {
```
Comment (jantimon) [inline lib/compiler.js:333]: why did you remove this part?

File: spec/caching.spec.js (hunk 6 above)
Comment (fivethreeo) [reply spec/caching.spec.js:142]: Yes :)

**Comment by fivethreeo** [comment]
use the webpack5-dev branch it is without actions

File: spec/hot.spec.js (hunk 7 above)
Comment (fivethreeo) [reply spec/hot.spec.js:88]: True, I'll take it out :)

File: lib/compiler.js (hunk 8 above)
Comment (fivethreeo) [reply lib/compiler.js]: Comes from PutziSan's changes, I'll need to take a look.

Comment (fivethreeo) [reply lib/compiler.js]: It is because the checking went from being sync to async

File: lib/compiler.js (hunk 3 above)
Comment (fivethreeo) [reply lib/compiler.js]: Use a weakmap? or a weakmap on compiler?

File: package.json [hunk 9]
```diff
@@ -42,12 +43,12 @@
     "pug": "2.0.3",
     "pug-loader": "2.4.0",
     "rimraf": "2.6.3",
-    "semistandard": "13.0.1",
+    "semistandard": "^13.0.1",
```
Comment (jantimon) [inline package.json:46]: We should revert his part - otherwise unit tests might break randomly

Comment (fivethreeo) [reply package.json:46]: Oh, that snuck in. Will fix.

**Comment by fivethreeo** [comment]
But I need a decision from you. 
Typescript fails because of missing /wrong types if the wrong webpack in installed in dev.

There are four options. 

1. Ignore types in compiler-webpack5 and compiler-webpack4.
2. Ignore types selectively by modifying tsconfig.json based on the webpack version found in pretest.
3. Release two versions of html-webpack-plugin, either by scoping it or doing a 5.x series, 4.x for webpack4. And drop webpack4 in 5.x.
4. Spend a long time creating custom types supporting both webpack versions or figure out how to do this some other way in typescript.

1 and 2 are not optimal solutions and will probably cause issues down the line. 4 needs a typescript expert which I’m not :(. But rolling releases I can do. :)

**Comment by fivethreeo** [comment]
Close this and use #1392

File: lib/compiler.js (hunk 8 above)
Comment (fivethreeo) [reply lib/compiler.js]: Oh, I saw wrong. isChildCompilerCacheOutdated needs to go back in.
